### PR TITLE
Add intermediate headers support for update-client and refactor chain and prover interfaces

### PR DESCRIPTION
**Pull request by bluele**
- refactor some interfaces
  - LightClient: remove unnecessary methods and clarify method names(especially clarify on finalisation for header)
  - ChainI/ProverI: introduce QueryContext that keeps proof height and wraps `context.Context`
- fix https://github.com/hyperledger-labs/yui-relayer/issues/82
- [reorder proof generation and header setup](https://github.com/hyperledger-labs/yui-relayer/pull/83/commits/f3e3d308e42f5e7e3ef23f63e7eb98b6147f27ef)
  - it allows to remove unnecessary sync steps during packet relay with LCP
  - ref. https://github.com/datachainlab/lcp/pull/55
- [remove 'I' suffix from interface names](https://github.com/hyperledger-labs/yui-relayer/pull/83/commits/07ad300cb601f5d4268ef3c11061cdb0f53e8191)

### CODE REVIEW

File: core/provers.go [hunk 1]
```diff
@@ -20,36 +23,48 @@ type ProverI interface {
...
+	// GetLatestFinalizedHeader returns the latest finalized header on this chain
+	// The returned header is expected to be the latest one of headers that can be verified by the light client
+	GetLatestFinalizedHeader() (latestFinalizedHeader Header, err error)
 
-	// UpdateLightWithHeader updates a header on the light client and returns the header and height corresponding to the chain
-	UpdateLightWithHeader() (header HeaderI, provableHeight int64, queryableHeight int64, err error)
+	// SetupHeadersForUpdate returns the finalized header and any intermediate headers needed to apply it to the client on the counterpaty chain
+	// The order of the returned header slice should be as: [<intermediate headers>..., <update header>]
+	// if the header slice's length == 0 and err == nil, the relayer should skips the update-client
+	SetupHeadersForUpdate(dstChain ChainInfoICS02Querier, latestFinalizedHeader Header) ([]Header, error)
```
Comment (siburu) [inline core/provers.go:63]: Can we change `LightClient` interface as below?
- `GetLatestFinalizedHeader` -> `GetLatestFinalizedHeight`
- The second argument of `SetupHeadersForUpdate`: `Header` -> `Height`

If we change codes like this, in the case where we already have a target height, we can directly use it for `SetupHeadersForUpdate` without calling `GetLatestFinalizedHeader`.

File: core/naive-strategy.go [hunk 2]
```diff
@@ -294,18 +297,17 @@ func (st NaiveStrategy) UnrelayedAcknowledgements(src, dst *ProvableChain, sh Sy
 }
 
 // TODO add packet-timeout support
-func relayPackets(chain *ProvableChain, seqs []uint64, sh SyncHeadersI, sender sdk.AccAddress) ([]sdk.Msg, error) {
+func relayPackets(ctx QueryContext, chain *ProvableChain, seqs []uint64, sender sdk.AccAddress) ([]sdk.Msg, error) {
```
Comment (3100) [inline core/naive-strategy.go:300]: The names `relayPackets` and `relayAcks` seem not intuitive.
How about using different names, such as `collectPackets` and `collectAcks`?

File: core/provers.go (hunk 1 above)
Comment (bluele) [reply core/provers.go:63]: I understand that your suggestion assumes that the past finalised headers can be retrieved using height at any point in time, it is somewhat difficult to satisfy this in ethereum at least.

In ethereum's sync protocol, it is difficult to get a finalised header corresponding to any height via the standard API: ref. https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/p2p-interface.md#light_client_finality_update

> in the case where we already have a target height

Although `GetLatestFinalisedHeader` must be called, we can avoid unnecessary (and expensive) update-client(s) if `SetupHeadersForUpdate` returns empty.

File: core/headers.go [hunk 3]
```diff
@@ -1,94 +1,126 @@
...
+func (sh syncHeaders) SetupHeadersForUpdate(src, dst ChainICS02QuerierLightClient) ([]Header, error) {
+	if err := ensureDifferentChains(src, dst); err != nil {
+		return nil, err
+	}
+	return src.SetupHeadersForUpdate(dst, sh.GetLatestFinalizedHeader(src.ChainID()))
+}
+
+// SetupBothHeadersForUpdate returns both `src` and `dst` chain's headers to update the clients on each chain
+func (sh syncHeaders) SetupBothHeadersForUpdate(src, dst ChainICS02QuerierLightClient) ([]Header, []Header, error) {
+	if err := ensureDifferentChains(src, dst); err != nil {
```
Comment (bluele) [inline core/headers.go:106]: oh, this call is completely redundant. I'll remove this later.

File: core/headers.go [hunk 4]
```diff
@@ -1,94 +1,126 @@
...
+// SetupHeadersForUpdate returns `src` chain's headers to update the client on `dst` chain
+func (sh syncHeaders) SetupHeadersForUpdate(src, dst ChainICS02QuerierLightClient) ([]Header, error) {
+	if err := ensureDifferentChains(src, dst); err != nil {
+		return nil, err
+	}
+	return src.SetupHeadersForUpdate(dst, sh.GetLatestFinalizedHeader(src.ChainID()))
+}
+
+// SetupBothHeadersForUpdate returns both `src` and `dst` chain's headers to update the clients on each chain
+func (sh syncHeaders) SetupBothHeadersForUpdate(src, dst ChainICS02QuerierLightClient) ([]Header, []Header, error) {
```
[the remark this answers is not in the document]
Comment (bluele) [reply core/headers.go:105]: https://github.com/hyperledger-labs/yui-relayer/pull/83/files/5e603ff0fd1de89bd9ebf79129196fa143462402#r1133902556

**Review by 3100 — APPROVED** [review]

File: chains/tendermint/prover.go [hunk 5]
```diff
@@ -161,50 +129,72 @@ func (pr *Prover) SetupHeader(dstChain core.LightClientIBCQueryierI, srcHeader c
...
+
+	client, err := pr.LightClient(db)
+	if err != nil {
+		return -1, err
+	}
+
+	return client.LastTrustedHeight()
+}
+
+func (pr *Prover) UpdateLightClient() (core.Header, error) {
```
Comment (siburu) [inline chains/tendermint/prover.go:162]: You can unexport this function.

File: chains/tendermint/prover.go [hunk 6]
```diff
@@ -161,50 +129,72 @@ func (pr *Prover) SetupHeader(dstChain core.LightClientIBCQueryierI, srcHeader c
...
+	}
+	return h, nil
+}
 
-// UpdateLightWithHeader calls client.Update and then .
-func (pr *Prover) UpdateLightWithHeader() (header core.HeaderI, provableHeight int64, queryableHeight int64, err error) {
+/* Local LightClient implementation */
+
+// GetLatestLightHeight uses the CLI utilities to pull the latest height from a given chain
+func (pr *Prover) GetLatestLightHeight() (int64, error) {
```
Comment (siburu) [inline chains/tendermint/prover.go:147]: Can you unexport this function?

File: cmd/query.go [hunk 7]
```diff
@@ -46,22 +47,16 @@ func queryClientCmd(ctx *config.Context) *cobra.Command {
 			}
 			c := chains[args[1]]
 
-			height, err := cmd.Flags().GetInt64(flags.FlagHeight)
+			height, err := c.LatestHeight()
```
Comment (siburu) [inline cmd/query.go:50]: Was the original code that referenced the `height` flag wrong?

File: core/naive-strategy.go [hunk 8]
```diff
@@ -294,18 +297,17 @@ func (st NaiveStrategy) UnrelayedAcknowledgements(src, dst *ProvableChain, sh Sy
 }
 
 // TODO add packet-timeout support
-func relayPackets(chain *ProvableChain, seqs []uint64, sh SyncHeadersI, sender sdk.AccAddress) ([]sdk.Msg, error) {
+func collectPackets(ctx QueryContext, chain *ProvableChain, seqs []uint64, sender sdk.AccAddress) ([]sdk.Msg, error) {
```
Comment (siburu) [inline core/naive-strategy.go:300]: `sender` can be mistaken for a packet sender.
Shouldn't it be renamed to `signer` or `txSender`?

File: core/naive-strategy.go [hunk 9]
```diff
@@ -328,84 +330,81 @@ func (st NaiveStrategy) RelayAcknowledgements(src, dst *ProvableChain, sp *Relay
...
+		if num := len(acksForSrc); num > 0 {
+			logPacketsRelayed(src, dst, num)
 		}
 	}
 
 	return nil
 }
 
-func relayAcks(receiverChain, senderChain *ProvableChain, seqs []uint64, sh SyncHeadersI, sender sdk.AccAddress) ([]sdk.Msg, error) {
+func collectAcks(senderCtx, receriverCtx QueryContext, senderChain, receiverChain *ProvableChain, seqs []uint64, sender sdk.AccAddress) ([]sdk.Msg, error) {
```
Comment (siburu) [inline core/naive-strategy.go:395]: typo: receiverCtx

File: chains/tendermint/prover.go (hunk 6 above)
Comment (bluele) [reply chains/tendermint/prover.go:147]: https://github.com/hyperledger-labs/yui-relayer/blob/07ad300cb601f5d4268ef3c11061cdb0f53e8191/chains/tendermint/cmd/light.go#L151
It is referenced in the option of the `header` command

File: core/naive-strategy.go (hunk 9 above)
Comment (bluele) [reply core/naive-strategy.go]: fixed at https://github.com/hyperledger-labs/yui-relayer/pull/83/commits/6881057cd3b323e434e5660e9e8a208ff4769388

File: core/naive-strategy.go (hunk 8 above)
Comment (bluele) [reply core/naive-strategy.go]: fixed at https://github.com/hyperledger-labs/yui-relayer/pull/83/commits/6881057cd3b323e434e5660e9e8a208ff4769388

File: cmd/query.go (hunk 7 above)
Comment (bluele) [reply cmd/query.go]: No, I've reverted the change and implemented the same option for other queriers(connection, channel) at https://github.com/hyperledger-labs/yui-relayer/pull/83/commits/64f72896878ee712ec823c27d77181b0ed08e27f

File: chains/tendermint/prover.go (hunk 5 above)
Comment (bluele) [reply chains/tendermint/prover.go:162]: https://github.com/hyperledger-labs/yui-relayer/blob/07ad300cb601f5d4268ef3c11061cdb0f53e8191/chains/tendermint/cmd/light.go#L106
It is also referenced in the option of the `update` command.

**Review by siburu — APPROVED** [review]
LGTM

**Pull request merged**